crates/matrix-sdk-appservice/src/lib.rs

```diff
@@ -627,6 +632,89 @@ impl AppService {
+    pub async fn receive_transaction(
+        &self,
+        transaction: push_events::v1::IncomingRequest,
+    ) -> Result<()> {
+        let client = self.get_cached_client(None)?;
+
+        // Find membership events affecting members in our namespace, and update
+        // membership accordingly
+        for event in transaction.events.iter() {
+            let event = match event.deserialize() {
+                Ok(AnyRoomEvent::State(AnyStateEvent::RoomMember(event))) => event,
+                _ => continue,
+            };
+            if !self.user_id_is_in_namespace(event.state_key())? {
+                continue;
+            }
+            let localpart = event.state_key().localpart();
+            client
+                .store()
+                .set_custom_value(
+                    &[USER_MEMBER, event.room_id().as_bytes(), localpart.as_bytes(), b"."].concat(),
+                    event.membership().to_string().into_bytes(),
+                )
+                .await?;
+        }
+
+        #[derive(Debug, Deserialize)]
+        struct EventRoomId {
+            room_id: Option<OwnedRoomId>,
+        }
+
+        // Send a sync event to each active client
+        // TODO: What do we do with state events?
+        for virt_client in self.clients.iter() {
+            let user_id = match virt_client.user_id() {
+                Some(user_id) => user_id,
+                // Unauthenticated client. (should that be possible?)
+                None => continue,
+            };
+            let mut response = sync_events::v3::Response::new(transaction.txn_id.to_string());
+
+            for raw_event in &transaction.events {
+                let room_id = match raw_event.deserialize_as::<EventRoomId>().unwrap().room_id {
+                    Some(room_id) => room_id,
+                    None => {
+                        warn!("Transaction contained event with no ID");
+                        continue;
+                    }
+                };
+                let key = &[USER_MEMBER, room_id.as_bytes(), b".", user_id.as_bytes()].concat();
+                let membership = match client.store().get_custom_value(key).await? {
+                    Some(value) => String::from_utf8(value).ok().map(MembershipState::from),
+                    None => None,
+                };
+                let json = Raw::into_json(raw_event.clone());
+
+                match membership.unwrap_or(MembershipState::Join) {
+                    MembershipState::Join => {
+                        let room = response.rooms.join.entry(room_id).or_default();
+                        room.timeline.events.push(Raw::from_json(json))
+                    }
+                    MembershipState::Leave | MembershipState::Ban => {
+                        let room = response.rooms.leave.entry(room_id).or_default();
+                        room.timeline.events.push(Raw::from_json(json))
```

Comment:
```suggestion

                match membership.unwrap_or(MembershipState::Join) {
                    MembershipState::Join => {
                        let room = response.rooms.join.entry(room_id).or_default();
                        room.timeline.events.push(raw_event.clone().cast());
                    }
                    MembershipState::Leave | MembershipState::Ban => {
                        let room = response.rooms.leave.entry(room_id).or_default();
                        room.timeline.events.push(raw_event.clone().cast());
```

Reply:
Thanks, I felt like there had to be a better way of doing that.